core/src/main/java/io/questdb/griffin/engine/functions/groupby/CountDistinctIntGroupByFunction.java

```diff
@@ -32,61 +32,57 @@
 import io.questdb.std.Numbers;
-import io.questdb.std.ObjList;
 
 public class CountDistinctIntGroupByFunction extends LongFunction implements UnaryFunction, GroupByFunction {
     private final Function arg;
-    private final int setInitialCapacity;
-    private final double setLoadFactor;
-    private final ObjList<CompactIntHashSet> sets = new ObjList<>();
-    private int setIndex;
+    private final GroupByIntHashSet setA;
+    private final GroupByIntHashSet setB;
     private int valueIndex;
 
     public CountDistinctIntGroupByFunction(Function arg, int setInitialCapacity, double setLoadFactor) {
         this.arg = arg;
-        this.setInitialCapacity = setInitialCapacity;
-        this.setLoadFactor = setLoadFactor;
+        // We use zero as the default value to speed up zeroing on rehash.
+        setA = new GroupByIntHashSet(setInitialCapacity, setLoadFactor, 0);
+        setB = new GroupByIntHashSet(setInitialCapacity, setLoadFactor, 0);
     }
 
     @Override
     public void clear() {
-        sets.clear();
-        setIndex = 0;
+        setA.resetPtr();
+        setB.resetPtr();
     }
 
     @Override
     public void computeFirst(MapValue mapValue, Record record) {
-        final CompactIntHashSet set;
-        if (sets.size() <= setIndex) {
-            sets.extendAndSet(setIndex, set = new CompactIntHashSet(setInitialCapacity, setLoadFactor, Numbers.INT_NaN));
-        } else {
-            set = sets.getQuick(setIndex);
-            set.clear();
-        }
-
-        final int val = arg.getInt(record);
+        int val = arg.getInt(record);
         if (val != Numbers.INT_NaN) {
-            set.add(val);
-            mapValue.putLong(valueIndex, 1L);
+            mapValue.putLong(valueIndex, 1);
+            // Remap zero since it's used as the no entry key.
+            val = (val == 0) ? Numbers.INT_NaN : val;
+            setA.of(0).add(val);
+            mapValue.putLong(valueIndex + 1, setA.ptr());
         } else {
-            mapValue.putLong(valueIndex, 0L);
+            mapValue.putLong(valueIndex, 0);
+            mapValue.putLong(valueIndex + 1, 0);
         }
-        mapValue.putInt(valueIndex + 1, setIndex++);
     }
 
     @Override
     public void computeNext(MapValue mapValue, Record record) {
-        final CompactIntHashSet set = sets.getQuick(mapValue.getInt(valueIndex + 1));
-        final int val = arg.getInt(record);
+        int val = arg.getInt(record);
         if (val != Numbers.INT_NaN) {
-            final int index = set.keyIndex(val);
-            if (index < 0) {
-                return;
+            long ptr = mapValue.getLong(valueIndex + 1);
+            // Remap zero since it's used as the no entry key.
+            val = (val == 0) ? Numbers.INT_NaN : val;
+            final int index = setA.of(ptr).keyIndex(val);
+            if (index >= 0) {
+                setA.addAt(index, val);
+                mapValue.addLong(valueIndex, 1);
```

Comment:
As far as I can see here `valueIndex` slot contains a non-distinct count, see another comment referring to this.

Reply:
`if (index >= 0)` stands true only if there was no such value. When this happens, we increase the counter by 1. This is count distinct. Am I missing something?

Reply:
Ok, so it's a distinct count indeed